### ISSUE TITLE
Cannot access template ref inside of component function

### ISSUE THREAD

**Issue by google-mac** [vuejs/composition-api#224]
Based on the documentation [here](https://vue-composition-api-rfc.netlify.com/api.html#template-refs), I would expect to be able to access a ref's value and call `focus()` on it. I cannot because it is always null.

The API reference shows that it is defined in `onMounted`, but I cannot use that lifecycle method nested inside of the function or vice-versa. How can I do this?

Also, on the TypeScript side, it would be helpful to know what the Ref type should be - if I put `null` in, I cannot call DOM element functions on it. `Element` does not seem to be the right one.

My code looks something like this:

```vue
<template>
    <div>
        <button @click="editLabel">Edit</button>

        <input
            ref="newLabelInput"
            v-model="newLabel"
            @keyup.enter="updateLabel"
        />        
    </div>
</template>

<script lang="ts">
import { createComponent, ref, computed } from '@vue/composition-api';

export default createComponent({
    setup({ file }) {
        const newLabel = ref<string>(file.label);
        const newLabelInput = ref<Element>(null);

        function editLabel() {
            console.log(newLabelInput.value); // always null
            newLabelInput.value.focus();
        }

        function updateLabel() {
            console.log(newLabelInput.value); // always null
            newLabelInput.value.focus();
        }

        return {
            newLabel,
            newLabelInput,
            editLabel,
            updateLabel,
        };
    },

    props: {
        file: {
            type: Object,
            required: true,
        },
    },
});
</script>
```

**Comment by google-mac** [comment]
I feel like this (DOM manipulation on user-triggered functions) is a fairly large use case for template refs. It should certainly be solved for Vue 3, though I would assume there's some way to do it for this composition API.

**Comment by google-mac** [comment]
I was able to figure it out, but it was only by placing the correct TS Type, like so:

```ts
const newLabelInput = ref<HTMLElement>(null);
```

Obviously, `Element` was not correct here, but I think the issue is now that it is not documented AND that if no type is placed or the incorrect one is, it actually affects the runtime and it does not work.

**Comment by posva** [comment]
I don't see how the type could affect the runtime.
Unrelated note: never destructure props in setup or reactivity is lost

**Comment by google-mac** [comment]
Yeah, I don't get it either, but that's what I've been banging my head against. I'd love to figure out why.

@posva In regards to using props in setup, it was only for a default value.

**Comment by google-mac** [comment]
I just tried it in a different project and it doesn't work, so the issue is still valid. There does not appear to be a way to interact with the ref in a user-triggered function (basically outside of a lifecycle method).

**Comment by LinusBorg** [comment]
I'm right now working on a project around keyboard a11y and I use template refs to focus elements all over the place - in Typescript.

Like posva, I can't follow your explanations fully.

The initial problem was, as it appears to me, that you used the `Element` type, which doesn't have a `focus` element, so TS would likely complain about that.

You however said it was "always `null`", which and that it wasn't when you changed the type. but Typescript doesn't work like that. Types don't affect runtime values like that.

Now you say "this is still an issue" but I simply can't see what *"this"* is.

Pleas provide a runnable reproduction.

**Comment by google-mac** [comment]
I just figured out what the issue is! So I have two projects that are using input focus. It works just fine in the first one, but not the second. I looked at them side by side until I found what could be the issue, and it hit me - I have a `v-if` on the second case's input field. The function that makes the `v-if` resolve to `true` immediately follows with attempting to focus on the input, which is essentially a race condition because the input was not always available (`null`) but sometimes it was.

In my case I can just replace `v-if` with `v-show`, but others that run across this may not be able to. After reading the docs on [template refs](https://vuejs.org/v2/api/#ref), my suggestion would be to also include this scenario where anything inside of a `v-if` will not be immediately available. Would this be where awaiting `Vue.nextTick()` would suffice before attempting to focus on the input?

PS. Yeah I totally get that TypeScript doesn't work that way. At the time, it was the only thing showing me any errors / output.
